File: src/services/ConversationAnalyzer.ts
import { Pool } from 'pg';

export class ConversationAnalyzer {
  private db: Pool;

  constructor() {
    this.db = new Pool({
      connectionString: process.env.DATABASE_URL || 'postgresql://localhost:5432/lending_sales'
    });
  }

  async analyzeObjections(objectionType?: string, timePeriod?: string) {
    let query = `
      SELECT objections, deal_stage, created_at 
      FROM conversations 
      WHERE objections != '[]' AND objections IS NOT NULL
    `;
    
    if (timePeriod) {
      query += ` AND created_at >= NOW() - INTERVAL '${timePeriod}'`;
    }
    
    const result = await this.db.query(query);
    const conversations = result.rows;
    
    // Flatten all objections
    const allObjections = conversations.flatMap(c => 
      JSON.parse(c.objections || '[]')
    );
    
    // Count frequency
    const objectionCounts = allObjections.reduce((acc, obj) => {
      acc[obj] = (acc[obj] || 0) + 1;
      return acc;
    }, {} as Record<string, number>);
    
    // Calculate success rates
    const successfulDeals = conversations.filter(c => 
      c.deal_stage === 'closed_won'
    ).length;
    
    const successRate = conversations.length > 0 
      ? Math.round((successfulDeals / conversations.length) * 100)
      : 0;
    
    return {
      mostCommon: Object.entries(objectionCounts)
        .sort(([,a], [,b]) => b - a)
        .slice(0, 5)
        .map(([objection]) => objection),
      successRate,
      recommendedResponses: this.getRecommendedResponses(objectionType)
    };
  }

  private getRecommendedResponses(objectionType?: string) {
    const responses = {
      'api_latency': [
        'Share performance benchmarks showing <2s response times',
        'Offer dedicated infrastructure for enterprise clients',
        'Provide SLA guarantees with penalty clauses'
      ],
      'pricing': [
        'Break down ROI calculations showing cost savings',
        'Offer tiered pricing based on volume',
        'Provide trial period with success-based pricing'
      ],
      'integration': [
        'Demonstrate plug-and-play API integration',
        'Offer dedicated technical support during integration',
        'Share integration timelines from similar clients'
      ]
    };
    
    return responses[objectionType as keyof typeof responses] || [
      'Acknowledge the concern and ask for specific requirements',
      'Provide relevant case studies and testimonials',
      'Offer a proof of concept or pilot program'
    ];
  }
} 

File: src/services/conversation_analyzer.py
from typing import Dict, Any, List
from datetime import datetime, timedelta
from pymongo import MongoClient
import os

class ConversationAnalyzer:
    def __init__(self, context: Dict[str, str]):
        self.context = context
        self.mongodb = MongoClient(os.getenv('MONGODB_URL'))
        self.conversations = self.mongodb.sales_memory.conversations

    async def analyze_objections(self, objection_type: str, time_period: str) -> Dict[str, Any]:
        # Calculate date range
        end_date = datetime.utcnow()
        if time_period == "last_week":
            start_date = end_date - timedelta(days=7)
        elif time_period == "last_month":
            start_date = end_date - timedelta(days=30)
        else:
            start_date = end_date - timedelta(days=90)

        # Get conversations with objections
        conversations = await self.conversations.find({
            "tenant_id": self.context["tenantId"],
            "created_at": {"$gte": start_date, "$lte": end_date},
            "objections": {"$exists": True, "$ne": []}
        }).to_list(length=None)

        # Flatten all objections
        all_objections = [obj for conv in conversations for obj in conv.get("objections", [])]
        
        # Count frequency
        objection_counts = {}
        for obj in all_objections:
            objection_counts[obj] = objection_counts.get(obj, 0) + 1
        
        # Calculate success rates
        successful_deals = len([c for c in conversations if c.get("deal_stage") == "closed_won"])
        success_rate = (successful_deals / len(conversations) * 100) if conversations else 0

        return {
            "most_common": sorted(
                objection_counts.items(),
                key=lambda x: x[1],
                reverse=True
            )[:5],
            "success_rate": round(success_rate, 1),
            "recommended_responses": self.generate_recommended_responses(objection_type)
        }

    def generate_recommended_responses(self, objection_type: str) -> List[str]:
        # This would typically come from your AI service
        return [
            "I understand your concern about pricing. Let me show you our ROI calculator...",
            "Many of our customers had similar concerns. Here's how they solved it...",
            "Let's break down the costs and benefits over a 12-month period..."
        ] 